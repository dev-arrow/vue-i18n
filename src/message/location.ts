import { isUnDef } from '../utils'

// The source range.
// The `start` is inclusive and `end` is exclusive.
// [start, end)
export interface SourceLocation {
  start: Position
  end: Position
  source?: string
}

export interface Position {
  offset: number // from start of content (file)
  line: number
  column: number
}

export function createPosition(line: number, column: number, offset: number) {
  return { line, column, offset } as Position
}

export function createLocation(
  start: Position,
  end: Position,
  source?: string
) {
  const loc = { start, end } as SourceLocation
  if (!isUnDef(source)) {
    loc.source = source
  }
  return loc
}
