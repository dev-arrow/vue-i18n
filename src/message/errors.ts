import { SourceLocation } from './location'
import { isObject } from '../utils'

export type CompileDomain =
  | 'tokenizer'
  | 'parser'
  | 'generator'
  | 'transformer'
  | 'compiler'

export interface CompileError extends SyntaxError {
  code: CompileErrorCodes
  domain?: CompileDomain
  location?: SourceLocation
}

export const enum CompileErrorCodes {
  // tokenizer error codes
  EXPECTED_TOKEN,
  INVALID_TOKEN_IN_PLACEHOLDER,
  UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER,
  UNKNOWN_ESCAPE_SEQUENCE,
  INVALID_UNICODE_ESCAPE_SEQUENCE,
  UNBALANCED_CLOSING_BRACE,
  UNTERMINATED_CLOSING_BRACE,
  EMPTY_PLACEHOLDER,
  NOT_ALLOW_NEST_PLACEHOLDER,
  INVALID_LINKED_FORMAT,

  // parser error codes
  MUST_HAVE_MESSAGES_IN_PLURAL,

  // Special value for higher-order compilers to pick up the last code
  // to avoid collision of error codes. This should always be kept as the last
  // item.
  __EXTEND_POINT__
}

export type CreateCompileErrorOptions = {
  domain?: CompileDomain
  messages?: { [code: number]: string }
  args?: unknown[]
}

export const errorMessages: { [code: number]: string } = {
  // tokenizer error messages
  [CompileErrorCodes.EXPECTED_TOKEN]: `Expected token: '{0}'`,
  [CompileErrorCodes.INVALID_TOKEN_IN_PLACEHOLDER]: `Invalid token in placeholder: '{0}'`,
  [CompileErrorCodes.UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER]: `Unterminated single quote in placeholder`,
  [CompileErrorCodes.UNKNOWN_ESCAPE_SEQUENCE]: `Unknown escape sequence: \\{0}`,
  [CompileErrorCodes.INVALID_UNICODE_ESCAPE_SEQUENCE]: `Invalid unicode escape sequence: {0}`,
  [CompileErrorCodes.UNBALANCED_CLOSING_BRACE]: `Unbalanced closing brace`,
  [CompileErrorCodes.UNTERMINATED_CLOSING_BRACE]: `Unterminated closing brace`,
  [CompileErrorCodes.EMPTY_PLACEHOLDER]: `Empty placeholder`,
  [CompileErrorCodes.NOT_ALLOW_NEST_PLACEHOLDER]: `Not allowed nest placeholder`,
  [CompileErrorCodes.INVALID_LINKED_FORMAT]: `Invalid linked format`,
  // parser error messages
  [CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL]: `Plural must have messages`
}

const RE_ARGS = /\{([0-9a-zA-Z]+)\}/g

/* eslint-disable @typescript-eslint/no-explicit-any */
export function format(message: string, ...args: any): string {
  if (args.length === 1 && isObject(args[0])) {
    args = args[0]
  }
  if (!args || !args.hasOwnProperty) {
    args = {}
  }
  return message.replace(
    RE_ARGS,
    (match: string, identifier: string): string => {
      return args.hasOwnProperty(identifier) ? args[identifier] : ''
    }
  )
}
/* eslint-enable @typescript-eslint/no-explicit-any */

export function createCompileError(
  code: CompileErrorCodes,
  loc: SourceLocation,
  optinos: CreateCompileErrorOptions = {}
): CompileError {
  const { domain, messages, args } = optinos
  const msg = __DEV__
    ? format((messages || errorMessages)[code] || '', ...(args || []))
    : code
  const error = new SyntaxError(String(msg)) as CompileError
  error.code = code
  error.location = loc
  error.domain = domain
  return error
}

export function defaultOnError(error: CompileError) {
  throw error
}
