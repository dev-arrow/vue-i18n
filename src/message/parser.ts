import { SourceLocation, Position } from './location'
import { createTokenizer, Tokenizer, TokenTypes, TokenizeContext } from './tokenizer'

export const enum NodeTypes {
  Resource, // 0
  Plural,
  Message,
  Text,
  Named,
  List, // 5
  Linked,
  LinkedKey,
  LinkedModifier
}

// not containing whitespace or control characters
export type Identifier = string

export interface Node {
  type: NodeTypes
  start: number
  end: number
  loc?: SourceLocation
}

export interface ResourceNode extends Node {
  type: NodeTypes.Resource
  body: MessageNode | PluralNode
}

export interface PluralNode extends Node {
  type: NodeTypes.Plural
  cases: MessageNode[]
}

export interface MessageNode extends Node {
  type: NodeTypes.Message
  items: MessageElementNode[]
}

type MessageElementNode =
  | TextNode
  | NamedNode
  | ListNode
  | LinkedNode

export interface TextNode extends Node {
  type: NodeTypes.Text
  value: string
}

export interface NamedNode extends Node {
  type: NodeTypes.Named
  key: Identifier
}

export interface ListNode extends Node {
  type: NodeTypes.List
  index: number
}

export interface LinkedNode extends Node {
  type: NodeTypes.Linked
  modifier?: LinkedModitierNode
  key: LinkedKeyNode | NamedNode | ListNode
}

export interface LinkedKeyNode extends Node {
  type: NodeTypes.LinkedKey
  value: string
}

export interface LinkedModitierNode extends Node {
  type: NodeTypes.LinkedModifier
  value: Identifier
}

type Parser = Readonly<{
  parse: (source: string) => ResourceNode
}>

export function createParser (): Parser {
  const startNode = (context: TokenizeContext, type: NodeTypes, init?: { offset: number, loc: Position }): Node => {
    let { lastOffset, lastStartLoc } = context
    if (init) {
      lastOffset = init.offset
      lastStartLoc = init.loc
    }
    return {
      type,
      start: lastOffset,
      end: -1,
      loc: { start: lastStartLoc, end: lastStartLoc }
    }
  }

  const endNode = (tokenizer: Tokenizer, node: Node, type?: NodeTypes): void => {
    node.end = tokenizer.currentOffset()
    if (type) {
      node.type = type
    }
    if (node.loc) {
      node.loc.end = tokenizer.currentPosition()
    }
  }

  const parseText = (tokenizer: Tokenizer, value: string): TextNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.Text) as TextNode
    node.value = value
    endNode(tokenizer, node)
    return node
  }

  const parseList = (tokenizer: Tokenizer, index: number): ListNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.List) as ListNode
    node.index = index
    // skip brach right
    tokenizer.nextToken()
    endNode(tokenizer, node)
    return node
  }

  const parseNamed = (tokenizer: Tokenizer, key: string): NamedNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.Named) as NamedNode
    node.key = key
    // skip brach right
    tokenizer.nextToken()
    endNode(tokenizer, node)
    return node
  }

  const parseLinkedModifier = (tokenizer: Tokenizer): LinkedModitierNode => {
    const token = tokenizer.nextToken()
    // asset check token
    if (!token.value || typeof token.value === 'number') {
      // TODO: should be thrown syntax error
      throw new Error()
    }
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.LinkedModifier) as LinkedModitierNode
    node.value = token.value
    endNode(tokenizer, node)
    return node
  }

  const parseLinkedKey = (tokenizer: Tokenizer, value: string): LinkedKeyNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.LinkedKey) as LinkedKeyNode
    node.value = value
    endNode(tokenizer, node)
    return node
  }

  const parseLinked = (tokenizer: Tokenizer): LinkedNode => {
    const context = tokenizer.context()
    const linkedNode = startNode(context, NodeTypes.Linked) as LinkedNode

    let token = tokenizer.nextToken()
    if (token.type === TokenTypes.LinkedDot) {
      linkedNode.modifier = parseLinkedModifier(tokenizer)
      token = tokenizer.nextToken()
    }

    // asset check token
    if (token.type !== TokenTypes.LinkedDelimiter) {
      // TODO: should be thrown syntax error
      throw new Error()
    }
    token = tokenizer.nextToken()

    // skip paren left
    if (token.type === TokenTypes.ParenLeft) {
      token = tokenizer.nextToken()
    }

    // skip brace left
    if (token.type === TokenTypes.BraceLeft) {
      token = tokenizer.nextToken()
    }

    switch (token.type) {
      case TokenTypes.LinkedKey:
        if (!token.value || typeof token.value !== 'string') {
          // TODO: should be thrown syntax error
          throw new Error()
        }
        linkedNode.key = parseLinkedKey(tokenizer, token.value)
        break
      case TokenTypes.Named:
        if (!token.value || typeof token.value === 'number') {
          // TODO: should be thrown syntax error
          throw new Error()
        }
        linkedNode.key = parseNamed(tokenizer, token.value)
        break
      case TokenTypes.List:
        if (token.value === undefined || typeof token.value === 'string') {
          // TODO: should be thrown syntax error
          throw new Error()
        }
        linkedNode.key = parseList(tokenizer, token.value)
        break
      default:
        // TODO: should be thrown syntax error
        throw new Error()
    }

    endNode(tokenizer, linkedNode)
    return linkedNode
  }

  const parseMessage = (tokenizer: Tokenizer): MessageNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.Message) as MessageNode
    node.items = []

    do {
      const token = tokenizer.nextToken()
      switch (token.type) {
        case TokenTypes.Text:
          if (!token.value || typeof token.value === 'number') {
            // TODO: should be thrown syntax error
            throw new Error()
          }
          node.items.push(parseText(tokenizer, token.value))
          break
        case TokenTypes.List:
          if (token.value === undefined || typeof token.value === 'string') {
            // TODO: should be thrown syntax error
            throw new Error()
          }
          node.items.push(parseList(tokenizer, token.value))
          break
        case TokenTypes.Named:
          if (!token.value || typeof token.value === 'number') {
            // TODO: should be thrown syntax error
            throw new Error()
          }
          node.items.push(parseNamed(tokenizer, token.value))
          break
        case TokenTypes.LinkedAlias:
          node.items.push(parseLinked(tokenizer))
          break
        default:
          break
      }
    } while (context.currentType !== TokenTypes.EOF && context.currentType !== TokenTypes.Pipe)

    endNode(tokenizer, node)
    return node
  }

  const parsePlural = (tokenizer: Tokenizer, offset: number, loc: Position, msgNode: MessageNode): PluralNode => {
    const context = tokenizer.context()

    const node = startNode(context, NodeTypes.Plural, { offset, loc }) as PluralNode
    node.cases = []
    node.cases.push(msgNode)

    do {
      const msg = parseMessage(tokenizer)
      node.cases.push(msg)
    } while (context.currentType !== TokenTypes.EOF)

    endNode(tokenizer, node)
    return node
  }

  const parseResource = (tokenizer: Tokenizer): MessageNode | PluralNode => {
    const context = tokenizer.context()
    const { offset, startLoc } = context

    const msgNode = parseMessage(tokenizer)
    if (context.currentType === TokenTypes.EOF) {
      return msgNode
    } else {
      return parsePlural(tokenizer, offset, startLoc, msgNode)
    }
  }

  function parse (source: string): ResourceNode {
    const tokenizer = createTokenizer(source)
    const context = tokenizer.context()

    const node = startNode(context, NodeTypes.Resource) as ResourceNode
    node.body = parseResource(tokenizer)

    // assert wheather achieved to EOF
    if (context.currentType !== TokenTypes.EOF) {
      // TODO: should be thrown syntax error
      throw new Error()
    }

    endNode(tokenizer, node)
    return node
  }

  return Object.freeze({
    parse
  })
}
