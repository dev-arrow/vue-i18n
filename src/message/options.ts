import { CompileError } from './errors'

export type CompileErrorHandler = (error: CompileError) => void
export type CompileCacheKeyHandler = (source: string) => string

export type TokenizeOptions = {
  location?: boolean
  onError?: CompileErrorHandler
}

export type ParserOptions = {
  // Generate source map?
  // - Default: false
  // sourceMap?: boolean
  // Filename for source map generation.
  // - Default: `message.intl`
  // filename?: string
  location?: boolean
  onError?: CompileErrorHandler
}

export type TransformOptions = {
  onError?: CompileErrorHandler
}

export type CodeGenOptions = {
  mode?: 'normal' | 'arrow' // default normal
  onError?: CompileErrorHandler
}

export type CompileOptions = {
  warnHtmlMessage?: boolean
  onCacheKey?: CompileCacheKeyHandler
} & TransformOptions &
  CodeGenOptions &
  ParserOptions &
  TokenizeOptions
