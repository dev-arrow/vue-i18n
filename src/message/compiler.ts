import { createParser, ResourceNode, Node, NodeTypes } from './parser'

type CodeGenContext = {
  source?: string
  code: string
  indentLevel: number
  // line: number
  // column: number
  // offset: number
  // map?: SourceMapGenerator
}

type CodeGenerator = Readonly<{
  context: () => CodeGenContext
  push: (code: string) => void
  indent: () => void
  deindent: (withoutNewLine?: boolean) => void
  newline: () => void
}>

function createCodeGenerator (source?: string): CodeGenerator {
  const _context = {
    source,
    code: '',
    indentLevel: 0
  } as CodeGenContext

  const context = (): CodeGenContext => _context

  const push = (code: string): void => {
    _context.code += code
  }

  const _newline = (n: number): void => {
    push('\n' + `  `.repeat(n))
  }

  const indent = (): void => {
    _newline(++_context.indentLevel)
  }

  const deindent = (withoutNewLine?: boolean): void => {
    if (withoutNewLine) {
      --_context.indentLevel
    } else {
      _newline(--_context.indentLevel)
    }
  }

  const newline = (): void => {
    _newline(_context.indentLevel)
  }

  return Object.freeze({
    context, push, indent, deindent, newline
  })
}

export type CompileOptions = {
  // TODO: should be defined some options
  // Generate source map?
  // - Default: false
  // sourceMap?: boolean
  // Filename for source map generation.
  // - Default: `message.intl`
  // filename?: string
}

export type CompileResult = {
  code: string
  ast: ResourceNode
  // TODO: should be implemetend sourcemap
  // map?: RawSourceMap
}

export type Compiler = Readonly<{
  compile: (source: string, options?: CompileOptions) => CompileResult
}>

export function createCompiler (): Compiler {
  const _parser = createParser()

  // transform AST
  const transform = (ast: ResourceNode, options: CompileOptions): void => {
    // TODO:
  }

  const generateNode = (generator: CodeGenerator, node: Node): void => {
    generator.push(`''`)
    /*
    switch (node.type) {
      default:
        // TODO: should be handled with error
        throw new Error(`unhandled codegen node type: ${node.type}`)
    }
    */
  }

  // generate code from AST
  const generate = (ast: ResourceNode, options: CompileOptions): string => {
    const generator = createCodeGenerator(ast.loc && ast.loc.source)
    generator.push(`function __msg__ (ctx) {`)
    generator.indent()
    generator.push('return ')
    generateNode(generator, ast)
    generator.deindent()
    generator.push(`}`)
    return generator.context().code
  }

  const compile = (source: string, options: CompileOptions = {}): CompileResult => {
    const ast = _parser.parse(source)
    transform(ast, options)
    const code = generate(ast, options)
    return { code, ast }
  }

  return Object.freeze({
    compile
  })
}
