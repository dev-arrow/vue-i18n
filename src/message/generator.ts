import { ResourceNode, Node, NodeTypes, PluralNode, MessageNode, TextNode, ListNode, NamedNode } from './parser'

export const INTERPOLATE_CODE = `const interpolate = val => { return val == null ? "" : Array.isArray(val) || ((Object.prototype.toString.call(val) === "[object Object]") && val.toString === Object.prototype.toString) ? JSON.stringify(val, null, 2) : String(val) }`

type CodeGenContext = {
  source?: string
  code: string
  indentLevel: number
  // line: number
  // column: number
  // offset: number
  // map?: SourceMapGenerator
}

type CodeGenerator = Readonly<{
  context: () => CodeGenContext
  push: (code: string) => void
  indent: () => void
  deindent: (withoutNewLine?: boolean) => void
  newline: () => void
}>

function createCodeGenerator (source?: string): CodeGenerator {
  const _context = {
    source,
    code: '',
    indentLevel: 0
  } as CodeGenContext

  const context = (): CodeGenContext => _context

  const push = (code: string): void => {
    _context.code += code
  }

  const _newline = (n: number): void => {
    push('\n' + `  `.repeat(n))
  }

  const indent = (): void => {
    _newline(++_context.indentLevel)
  }

  const deindent = (withoutNewLine?: boolean): void => {
    if (withoutNewLine) {
      --_context.indentLevel
    } else {
      _newline(--_context.indentLevel)
    }
  }

  const newline = (): void => {
    _newline(_context.indentLevel)
  }

  return Object.freeze({
    context, push, indent, deindent, newline
  })
}

function generateMessageNode (generator: CodeGenerator, node: MessageNode): void {
  if (node.items.length > 1) {
    generator.push('[')
    generator.indent()
    for (let i = 0; i < node.items.length; i++) {
      generateNode(generator, node.items[i])
      generator.push(', ')
    }
    generator.push('""')
    generator.deindent()
    generator.push('].join("")')
  } else {
    generateNode(generator, node.items[0])
  }
}

function generatePluralNode (generator: CodeGenerator, node: PluralNode): void {
  if (node.cases.length > 1) {
    generator.push('[')
    generator.indent()
    for (let i = 0; i < node.cases.length; i++) {
      generateNode(generator, node.cases[i])
      generator.push(', ')
    }
    generator.push('""')
    generator.deindent()
    generator.push('].join("")')
  }
}

function generateResource (generator: CodeGenerator, node: ResourceNode): void {
  if (node.body) {
    generateNode(generator, node.body)
  } else {
    generator.push('null')
  }
}

function generateNode (generator: CodeGenerator, node: Node): void {
  switch (node.type) {
    case NodeTypes.Resource:
      generateResource(generator, (node as ResourceNode))
      break
    case NodeTypes.Plural:
      generatePluralNode(generator, (node as PluralNode))
      break
    case NodeTypes.Message:
      generateMessageNode(generator, (node as MessageNode))
      break
    case NodeTypes.List:
      generator.push(`interpolate(ctx.list[${(node as ListNode).index}])`)
      break
    case NodeTypes.Named:
      generator.push(`interpolate(ctx.named.${(node as NamedNode).key})`)
      break
    case NodeTypes.Text:
      generator.push(JSON.stringify((node as TextNode).value))
      break
    default:
      // TODO: should be handled with error
      throw new Error(`unhandled codegen node type: ${node.type}`)
  }
}

// generate code from AST
export const generate = (ast: ResourceNode): string => {
  const generator = createCodeGenerator(ast.loc && ast.loc.source)
  generator.push(`function __msg__ (ctx) {`)
  generator.indent()
  if (ast.needInterpolate) {
    generator.push(INTERPOLATE_CODE)
    generator.newline()
  }
  generator.push(`return `)
  generateNode(generator, ast)
  generator.deindent()
  generator.push(`}`)
  return generator.context().code
}
