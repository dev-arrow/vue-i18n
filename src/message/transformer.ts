import { ResourceNode, Node, NodeTypes, PluralNode, MessageNode, LinkedNode, LinkedModitierNode, LinkedKeyNode } from './parser'

// TODO: if we offer custom transform for uses, should be defined TransformOptions type to here
// ex.
// type TransformOptions = {
// }

type TransformContext = {
  ast: ResourceNode
  modifiers: Set<string>
  refers: Set<string>
  needInterpolate: boolean
}

type Transformer = Readonly<{
  context: () => TransformContext
}>

function createTransformer (ast: ResourceNode/*, options: TransformOptions */): Transformer {
  const _context = {
    ast,
    modifiers: new Set(),
    refers: new Set(),
    needInterpolate: false
  } as TransformContext

  const context = (): TransformContext => _context

  return Object.freeze({
    context
  })
}

function traverseNodes (nodes: Node[], transformer: Transformer): void {
  for (let i = 0; i < nodes.length; i++) {
    traverseNode(nodes[i], transformer)
  }
}

function traverseNode (node: Node, transformer: Transformer): void {
  const context = transformer.context()

  // TODO: if we need pre-hook of transform, should be implemeted to here

  switch (node.type) {
    case NodeTypes.Plural:
      traverseNodes((node as PluralNode).cases, transformer)
      break
    case NodeTypes.Message:
      traverseNodes((node as MessageNode).items, transformer)
      break
    case NodeTypes.Linked:
      const linked = node as LinkedNode
      linked.modifier && traverseNode(linked.modifier, transformer)
      traverseNode(linked.key, transformer)
      break
    case NodeTypes.LinkedModifier:
      const modifier = node as LinkedModitierNode
      context.modifiers.add(modifier.value)
      break
    case NodeTypes.LinkedKey:
      const key = node as LinkedKeyNode
      context.refers.add(key.value)
      break
    case NodeTypes.List:
    case NodeTypes.Named:
      context.needInterpolate = true
      break
    default:
      // TODO:
      break
  }

  // TODO: if we need post-hook of transform, should be implemeted to here
}

// transform AST
export function transform (ast: ResourceNode/*, options: TransformOptions */): void {
  const transformer = createTransformer(ast)
  // traverse
  ast.body && traverseNode(ast.body, transformer)
  // set meta information
  const context = transformer.context()
  ast.modifiers = [...context.modifiers]
  ast.refers = [...context.refers]
  ast.needInterpolate = context.needInterpolate
}
