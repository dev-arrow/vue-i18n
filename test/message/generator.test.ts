import { createParser } from '../../src/message/parser'
import { transform } from '../../src/message/transformer'
import { generate } from '../../src/message/generator'

describe('text', () => {
  test('basic', () => {
    const parser = createParser()
    const ast = parser.parse('hello world')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return "hello world"`)
    expect(code).toMatchSnapshot()
  })

  test('multline', () => {
    const parser = createParser()
    const ast = parser.parse('hello\n world')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return "hello\\n world"`)
    expect(code).toMatchSnapshot()
  })
})

describe('list', () => {
  test('basic', () => {
    const parser = createParser()
    const ast = parser.parse('hi {0} !')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`"hi ", ctx._interpolate(ctx.list[0]), " !", ""`)
    expect(code).toMatch(`].join("")`)
    expect(code).toMatchSnapshot()
  })

  test('multiple', () => {
    const parser = createParser()
    const ast = parser.parse('{0} {1} !')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`ctx._interpolate(ctx.list[0]), " ", ctx._interpolate(ctx.list[1]), " !", ""`)
    expect(code).toMatch(`].join("")`)
    expect(code).toMatchSnapshot()
  })
})

describe('named', () => {
  test('basic', () => {
    const parser = createParser()
    const ast = parser.parse('hi {name} !')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`"hi ", ctx._interpolate(ctx.named["name"]), " !", ""`)
    expect(code).toMatch(`].join("")`)
    expect(code).toMatchSnapshot()
  })

  test('multiple', () => {
    const parser = createParser()
    const ast = parser.parse('{greeting} {name} !')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`ctx._interpolate(ctx.named["greeting"]), " ", ctx._interpolate(ctx.named["name"]), " !", ""`)
    expect(code).toMatch(`].join("")`)
    expect(code).toMatchSnapshot()
  })
})

describe('linked', () => {
  test('key', () => {
    const parser = createParser()
    const ast = parser.parse('hi @:name !')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`"hi ", ctx._resolveMsg("name")(ctx), " !", ""`)
    expect(code).toMatch(`].join("")`)
    expect(code).toMatchSnapshot()
  })

  test('list', () => {
    const parser = createParser()
    const ast = parser.parse('hi @:{0} !')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`"hi ", ctx._resolveMsg(ctx._interpolate(ctx.list[0]))(ctx), " !", ""`)
    expect(code).toMatch(`].join("")`)
    expect(code).toMatchSnapshot()
  })

  test('named', () => {
    const parser = createParser()
    const ast = parser.parse('hi @:{name} !')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`"hi ", ctx._resolveMsg(ctx._interpolate(ctx.named["name"]))(ctx), " !", ""`)
    expect(code).toMatch(`].join("")`)
    expect(code).toMatchSnapshot()
  })

  test('modifier', () => {
    const parser = createParser()
    const ast = parser.parse('hi @.upper:(name) !')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`"hi ", ctx._resolveModifier("upper")(ctx._resolveMsg("name")(ctx)), " !", ""`)
    expect(code).toMatch(`].join("")`)
    expect(code).toMatchSnapshot()
  })
})

describe('plural', () => {
  test('simple', () => {
    const parser = createParser()
    const ast = parser.parse('no apples | one apple  |  too much apples  ')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`"no apples", "one apple", "too much apples  ", ""`)
    expect(code).toMatch(`[ctx.plural.rule(ctx.plural.index, 3)]`)
    expect(code).toMatchSnapshot()
  })

  test('complex', () => {
    const parser = createParser()
    const ast = parser.parse('@.caml:(no apples) | {0} apple | {n}　apples')
    transform(ast)
    const code = generate(ast)
    expect(code).toMatch(`return [`)
    expect(code).toMatch(`ctx._resolveModifier("caml")(ctx._resolveMsg("no apples")(ctx)), [`)
    expect(code).toMatch(`ctx._interpolate(ctx.list[0]), " apple", ""`)
    expect(code).toMatch(`ctx._interpolate(ctx.named["n"]), "　apples", ""`) // eslint-disable-line
    expect(code).toMatch(`[ctx.plural.rule(ctx.plural.index, 3)]`)
    expect(code).toMatchSnapshot()
  })
})
