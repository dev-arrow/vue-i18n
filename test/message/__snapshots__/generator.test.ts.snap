// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linked key: hi @:name ! 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    \\"hi \\", ctx.message(\\"name\\")(ctx), \\" !\\"
  ])
}"
`;

exports[`linked list: hi @:{0} ! 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    \\"hi \\", ctx.message(ctx.interpolate(ctx.list(0)))(ctx), \\" !\\"
  ])
}"
`;

exports[`linked modifier: hi @.upper:(name) ! 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    \\"hi \\", ctx.modifier(\\"upper\\")(ctx.message(\\"name\\")(ctx)), \\" !\\"
  ])
}"
`;

exports[`linked named: hi @:{name} ! 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    \\"hi \\", ctx.message(ctx.interpolate(ctx.named(\\"name\\")))(ctx), \\" !\\"
  ])
}"
`;

exports[`list basic: hi {0} ! 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    \\"hi \\", ctx.interpolate(ctx.list(0)), \\" !\\"
  ])
}"
`;

exports[`list multiple: {0} {1} ! 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    ctx.interpolate(ctx.list(0)), \\" \\", ctx.interpolate(ctx.list(1)), \\" !\\"
  ])
}"
`;

exports[`named basic: hi {name} ! 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    \\"hi \\", ctx.interpolate(ctx.named(\\"name\\")), \\" !\\"
  ])
}"
`;

exports[`named multiple: {greeting} {name} ! 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    ctx.interpolate(ctx.named(\\"greeting\\")), \\" \\", ctx.interpolate(ctx.named(\\"name\\")), \\" !\\"
  ])
}"
`;

exports[`plural complex: @.caml:(no apples) | {0} apple | {n}　apples 1`] = `
"function __msg__ (ctx) {
  return [
    ctx.process([
      ctx.modifier(\\"caml\\")(ctx.message(\\"no apples\\")(ctx))
    ]), ctx.process([
      ctx.interpolate(ctx.list(0)), \\" apple\\"
    ]), ctx.process([
      ctx.interpolate(ctx.named(\\"n\\")), \\"　apples\\"
    ])
  ][ctx.pluralRule(ctx.pluralIndex, 3, ctx.orgPluralRule)]
}"
`;

exports[`plural simple: no apples | one apple  |  too much apples   1`] = `
"function __msg__ (ctx) {
  return [
    ctx.process([
      \\"no apples\\"
    ]), ctx.process([
      \\"one apple\\"
    ]), ctx.process([
      \\"too much apples  \\"
    ])
  ][ctx.pluralRule(ctx.pluralIndex, 3, ctx.orgPluralRule)]
}"
`;

exports[`text basic: hello world 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    \\"hello world\\"
  ])
}"
`;

exports[`text multline: hello
 world 1`] = `
"function __msg__ (ctx) {
  return ctx.process([
    \\"hello\\\\n world\\"
  ])
}"
`;
